File: src/Tracking/TrackedEvent.php
<?php

declare(strict_types=1);

namespace App\Tracking;

/**
 * One visitor event as it arrives from the tracking script.
 *
 * The idempotency id is mandatory: k.js retries on a flaky connection, and a retried
 * purchase must not show up twice in the stream or count twice in revenue.
 */
final readonly class TrackedEvent
{
    private const SUPPORTED_TYPES = [
        'pageview', 'add_to_cart', 'remove_from_cart', 'purchase', 'login', 'signup',
        'search', 'wishlist', 'cart_abandon', 'email_open', 'email_click', 'popup_shown', 'coupon_used',
    ];

    public function __construct(
        public string $idempotencyId,
        public string $type,
        public \DateTimeImmutable $occurredAt,
        public string $detail,
        public string $customerId,
        public string $customerName,
        public string $siteName,
    ) {
    }

    /**
     * @param array<string, mixed> $payload
     *
     * @throws InvalidEventPayload
     */
    public static function fromPayload(array $payload): self
    {
        $idempotencyId = self::readString($payload, 'idempotency_id');
        if ('' === $idempotencyId) {
            throw new InvalidEventPayload('Pole „idempotency_id” jest wymagane.');
        }

        $type = self::readString($payload, 'type');
        if (!\in_array($type, self::SUPPORTED_TYPES, true)) {
            throw new InvalidEventPayload(sprintf('Nieznany typ zdarzenia „%s”.', $type));
        }

        return new self(
            idempotencyId: $idempotencyId,
            type: $type,
            occurredAt: self::readMoment($payload),
            detail: self::readString($payload, 'detail'),
            customerId: self::readString($payload, 'customer_id'),
            customerName: self::readString($payload, 'customer_name'),
            siteName: self::readString($payload, 'site'),
        );
    }

    /**
     * @param array<string, mixed> $payload
     */
    private static function readString(array $payload, string $key): string
    {
        $value = $payload[$key] ?? '';

        return \is_string($value) ? trim($value) : '';
    }

    /**
     * @param array<string, mixed> $payload
     *
     * @throws InvalidEventPayload
     */
    private static function readMoment(array $payload): \DateTimeImmutable
    {
        $occurredAt = self::readString($payload, 'occurred_at');
        if ('' === $occurredAt) {
            return new \DateTimeImmutable();
        }

        try {
            return new \DateTimeImmutable($occurredAt);
        } catch (\DateMalformedStringException $exception) {
            throw new InvalidEventPayload('Pole „occurred_at” nie jest poprawną datą.', previous: $exception);
        }
    }
}

File: backend/src/main/java/click/kivvi/domain/tracking/TrackedEvent.java
package click.kivvi.domain.tracking;

import java.time.DateTimeException;
import java.time.Instant;
import java.time.LocalDateTime;
import java.time.OffsetDateTime;
import java.time.ZoneOffset;
import java.time.format.DateTimeParseException;
import java.util.Map;

/**
 * One visitor event as it arrives from the tracking script — ported from {@code
 * App\Tracking\TrackedEvent}.
 *
 * <p>The idempotency id is mandatory: the tracking script retries on a flaky connection, and a
 * retried purchase must not show up twice in the stream or count twice in revenue.
 */
public record TrackedEvent(
    String idempotencyId,
    EventType type,
    Instant occurredAt,
    String detail,
    String customerId,
    String customerName,
    String siteName) {

  /**
   * @throws InvalidEventPayload with the exact Polish message {@code CollectController} returns,
   *     matching {@code TrackedEvent::fromPayload} field for field.
   */
  public static TrackedEvent fromPayload(Map<String, Object> payload) {
    String idempotencyId = readString(payload, "idempotency_id");
    if (idempotencyId.isEmpty()) {
      throw new InvalidEventPayload("Pole „idempotency_id” jest wymagane.");
    }

    String typeCode = readString(payload, "type");
    EventType type =
        EventType.fromCode(typeCode)
            .orElseThrow(
                () -> new InvalidEventPayload("Nieznany typ zdarzenia „" + typeCode + "”."));

    return new TrackedEvent(
        idempotencyId,
        type,
        readMoment(payload),
        readString(payload, "detail"),
        readString(payload, "customer_id"),
        readString(payload, "customer_name"),
        readString(payload, "site"));
  }

  private static String readString(Map<String, Object> payload, String key) {
    Object value = payload.get(key);
    return value instanceof String text ? text.trim() : "";
  }

  private static Instant readMoment(Map<String, Object> payload) {
    String occurredAt = readString(payload, "occurred_at");
    if (occurredAt.isEmpty()) {
      return Instant.now();
    }
    try {
      return parseFlexible(occurredAt);
    } catch (DateTimeException exception) {
      throw new InvalidEventPayload("Pole „occurred_at” nie jest poprawną datą.", exception);
    }
  }

  /**
   * Accepts the ISO-8601 shapes a tracking script realistically sends (with a zone offset, or bare
   * local time assumed UTC) — mirrors {@code new \DateTimeImmutable($value)}'s permissive parsing
   * without adopting its full free-form grammar.
   */
  private static Instant parseFlexible(String value) {
    try {
      return Instant.parse(value);
    } catch (DateTimeParseException ignored) {
      // Fall through to the next shape.
    }
    try {
      return OffsetDateTime.parse(value).toInstant();
    } catch (DateTimeParseException ignored) {
      // Fall through to the next shape.
    }
    return LocalDateTime.parse(value).toInstant(ZoneOffset.UTC);
  }
}
